### PR TITLE
Implemented code to get details for the activities within a specific pipeline run

### LINKED ISSUE #12: Capture Databricks in-pipeline execution time

### PR DESCRIPTION
**Pull request by nsftwr**
image

### CODE REVIEW

File: src/CarbonIntensityTime.DataFactory/Models/PipelineRun.cs [hunk 1]
```diff
@@ -10,4 +12,19 @@ public class PipelineRun
     public int? Duration { get; init; }
     public string InvokedByName { get; init; } = string.Empty;
     public string InvokedByType { get; init; } = string.Empty;
+    public IList<PipelineActivity> Activities { get; init; }
+}
+
+public class PipelineActivity
```
Comment (dazfuller) [inline src/CarbonIntensityTime.DataFactory/Models/PipelineRun.cs]: This needs moving into it's own ile

File: src/CarbonIntensityTime.DataFactory/Models/PipelineRun.cs [hunk 2]
```diff
@@ -10,4 +12,19 @@ public class PipelineRun
...
+{
+    public string ActivityId { get; init; } = string.Empty;
+    public string Name { get; init; } = string.Empty;
+    public string Type { get; init; } = string.Empty;
+    public string Status { get; init; } = string.Empty;
+    public DateTime? Start { get; init; }
+    public DateTime? End { get; init; }
+    public int? Duration { get; init; }
+    public object Input { get; init; }
```
Comment (dazfuller) [inline src/CarbonIntensityTime.DataFactory/Models/PipelineRun.cs]: What is this type? Check to see if the input can be transformed into a Dictionary

File: src/CarbonIntensityTime.DataFactory/Models/PipelineRun.cs [hunk 3]
```diff
@@ -10,4 +12,19 @@ public class PipelineRun
     public int? Duration { get; init; }
     public string InvokedByName { get; init; } = string.Empty;
     public string InvokedByType { get; init; } = string.Empty;
+    public IList<PipelineActivity> Activities { get; init; }
```
Comment (dazfuller) [inline src/CarbonIntensityTime.DataFactory/Models/PipelineRun.cs]: add ` = new();` to the end here to remove the nullability warnings

File: src/CarbonIntensityTime.DataFactory/FactoryClient.cs [hunk 4]
```diff
@@ -55,14 +55,14 @@ public async Task<ICollection<PipelineRun>> ListPipelineRuns(int minutes)
...
+
         var client = await GetClient();
 
         _logger.LogDebug("Retrieving pipelines runs");
-        var x = (await client.PipelineRuns.QueryByFactoryAsync(_factorySettings.ResourceGroup, _factorySettings.Name,
+        var getPipelineRuns = (await client.PipelineRuns.QueryByFactoryAsync(_factorySettings.ResourceGroup, _factorySettings.Name,
             filter)).Value;
 
-        return x.Select(pr => new PipelineRun
+        var pipelineRuns = getPipelineRuns.Select(async pr => new PipelineRun
```
Comment (dazfuller) [inline src/CarbonIntensityTime.DataFactory/FactoryClient.cs:65]: Need to make sure we're not going to encounter 429 issues here when potentially reading a lot of pipelines and activities

File: README.md [hunk 5]
```diff
@@ -33,5 +34,6 @@ The intent of this solution is to be able to collect anonymous telemetry from al
     "ClientSecret": null
   }
 }
-
 ```
+
+Note: If no ClientId and ClientSecret is specified, then AzureDefaultCredential will be used.
```
Comment (dazfuller) [inline README.md]: Suggest changing this to

> If no client id or secret is specified then the application will attempt to use the [AzureDefaultCredential](https://learn.microsoft.com/dotnet/api/azure.identity.defaultazurecredential) type which will check for Azure CLI, Visual Studio Code, etc... type of credentials first. If you are logged into a tenant other than the one in which the Azure Data Factory instance exists then you may get an error when running the code.

**Review by dazfuller — CHANGES_REQUESTED** [review]

File: src/CarbonIntensityTime.DataFactory/FactoryClient.cs (hunk 4 above)
Comment (nsftwr) [reply src/CarbonIntensityTime.DataFactory/FactoryClient.cs:65]: we will need to review as part of the SDK migration

**Review by dazfuller — APPROVED** [review]

**Pull request merged**